src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,467 @@
+  predicate to its flat inductive:
+  ```
+    info: infSeq.functor_unfold (α : Type) (r : α → α → Prop) (a✝ : α) : infSeq α r a✝ = infSeq_functor α r (infSeq α r) a✝
+  ```
+
+  We use these to define all the constructors from the original definition. For example, we obtain:
+  ```
+    infSeq.step (α : Type) (r : α → α → Prop) {a b : α} : r a b → infSeq α r b → infSeq α r a
+  ```
+
+  Similarly, we obtain the associated `casesOn` lemma (that are generated by `mkCasesOnCoinductive`):
+  ```
+    infSeq.casesOn (α : Type) (r : α → α → Prop) {motive : (a : α) → infSeq α r a → Prop} {a✝ : α} (t : infSeq α r a✝)
+    (step : ∀ {a b : α} (a_1 : r a b) (a_2 : infSeq α r b), motive a (infSeq.step α r a_1 a_2)) : motive a✝ t
+  ```
+
+  At the very end, we make use of the syntax references from the original `InductiveView`s
+  and set them to newly generated constructors. We apply deriving handlers and docstrings.
+  Note that attributes and modifiers are handled earlier by `PartialFixpoint` machinery
+-/
+
+/-- This structure contains the data carried in `InductiveElabStep1` that are solely used in
+mutual coinductive predicate elaboration. -/
+public structure CoinductiveElabData where
+  /-- Declaration name of the predicate-/
+  declName : Name
+  /-- Ref from the original `InductiveView`-/
+  ref : Syntax
+  /-- Modifiers from the original `InductiveView`-/
+  modifiers : Modifiers
+  /-- Constructor refs from the original `InductiveView`-/
+  ctorSyntax : Array Syntax
+  /-- The flag that is `true` if the predicate was defined via `coinductive` keyword and `false`
+  otherwise. When we elaborate a mutual definition, we allow mixing `coinductive` and `inductive`
+  keywords, and hence we need to record this information.
+  -/
+  isGreatest : Bool
+  deriving Inhabited
+
+public def addFunctorPostfix : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str p (s ++ "_functor")
+  | Name.num p n => Name.num (addFunctorPostfix p) n
+
+public def removeFunctorPostfix : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str p (s.stripSuffix "_functor")
+  | Name.num p n => Name.num (removeFunctorPostfix p) n
+
+public def removeFunctorPostfixInCtor : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str (removeFunctorPostfix p) s
+  | Name.num p n => Name.num (removeFunctorPostfixInCtor p) n
+
+private def rewriteGoalUsingEq (goal : MVarId) (eq : Expr) (symm : Bool := false) : MetaM MVarId := do
+  let rewriteResult ← goal.rewrite (←goal.getType) eq symm
+  goal.replaceTargetEq rewriteResult.eNew rewriteResult.eqProof
+
+/--
+  Generates unrolling lemmas that relate coinductive predicates to their flat inductive forms.
+  These lemmas are essential for the constructor generation process, providing the bridge
+  between the user-facing coinductive predicates and their internal flat representations.
+-/
```

Comment:
This is the 
```
    info: infSeq.functor_unfold (α : Type) (r : α → α → Prop) (a✝ : α) : infSeq α r a✝ = infSeq_functor α r (infSeq α r) a✝
```
from above? The maybe call it “unfolding” not “unrolling” for consistent terminology. And maybe cite that running example type here.